actors/builtin/miner/miner_actor.go

```diff
@@ -332,33 +335,61 @@ func (a Actor) PreCommitSector(rt Runtime, params *SectorPreCommitInfo) *adt.Emp
 		// when it arrives.
 		rt.Abortf(exitcode.ErrIllegalArgument, "seal challenge epoch %v too old, must be after %v", params.SealRandEpoch, challengeEarliest)
 	}
+	if params.ReplaceSector != 0 && len(params.DealIDs) == 0 {
+		rt.Abortf(exitcode.ErrIllegalArgument, "cannot replace sector without committing deals")
+	}
 
 	store := adt.AsStore(rt)
 	var st State
 	newlyVestedAmount := rt.State().Transaction(&st, func() interface{} {
 		rt.ValidateImmediateCallerIs(st.Info.Worker)
 		if params.SealProof != st.Info.SealProofType {
-			rt.Abortf(exitcode.ErrIllegalArgument, "wrong proof type")
+			rt.Abortf(exitcode.ErrIllegalArgument, "sector seal proof %v must match miner seal proof type %d", params.SealProof, st.Info.SealProofType)
 		}
 
-		if _, found, err := st.GetPrecommittedSector(store, params.SectorNumber); err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "failed to check precommit %v: %v", params.SectorNumber, err)
-		} else if found {
-			// Sector is currently precommitted but still not proven.
-			rt.Abortf(exitcode.ErrIllegalArgument, "sector %v already precommitted", params.SectorNumber)
+		_, preCommitFound, err := st.GetPrecommittedSector(store, params.SectorNumber)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check pre-commit %v", params.SectorNumber)
+		if preCommitFound {
+			rt.Abortf(exitcode.ErrIllegalArgument, "sector %v already pre-committed", params.SectorNumber)
 		}
 
-		if found, err := st.HasSectorNo(store, params.SectorNumber); err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "failed to check sector %v: %v", params.SectorNumber, err)
-		} else if found {
+		sectorFound, err := st.HasSectorNo(store, params.SectorNumber)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check sector %v", params.SectorNumber)
+		if sectorFound {
 			rt.Abortf(exitcode.ErrIllegalArgument, "sector %v already committed", params.SectorNumber)
 		}
 
+		if params.ReplaceSector != 0 {
+			replaceSector, found, err := st.GetSector(store, params.ReplaceSector)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sector %v", params.SectorNumber)
+			if !found {
+				rt.Abortf(exitcode.ErrNotFound, "no such sector %v to replace", params.ReplaceSector)
+			}
+			if len(replaceSector.DealIDs) > 0 {
+				rt.Abortf(exitcode.ErrIllegalArgument, "cannot replace sector %v which has deals", params.ReplaceSector)
+			}
+			// The sectors must be the same size, but that is always true because each miner actor supports
+			// only one sector size.
+			if params.Expiration < replaceSector.Expiration {
+				rt.Abortf(exitcode.ErrIllegalArgument, "cannot replace sector %v expiration %v with sooner expiration %v",
+					params.ReplaceSector, replaceSector.Expiration, params.Expiration)
+			}
+			faulty, err := st.Faults.IsSet(uint64(params.ReplaceSector))
```

Comment:
This check should ensure that we are not in the two cases: the sector is faulty or that the sector is active but declared faulted.

Reply:
These are the same. However a sector ends up in a fault state, it's set here.